### ISSUE TITLE
Feature Request: add conditions to associations

### ISSUE THREAD

**Issue by atiertant** [balderdashy/waterline#988]
hi, i would like to add conditions to associations like this :

    var User = Waterline.Collection.extend({
      identity: 'user',
      connection: 'local-postgresql',
      attributes: {
        firstName: 'string',
        lastName: 'string',
        dog: {
          collection: 'pet',
          via: 'owners',
          conditions: {
            legs: 4 
          },
          dominant: true
        }
      }
    });

the association conditions whould be add to populate's one.

**Comment by dmarcelino** [comment]
Hi @atiertant, if understand correctly you want way to predefine populate queries through your models, so a `User.find().populate('dog')` would automatically become:
``` javascript
User.find().populate('dog', { legs: 4 })
```

Did I get it right?

**Comment by atiertant** [comment]
@dmarcelino that's it and User.find().populate('dog', {color: 'brown'}) automaticaly becomes :

    User.find().populate('dog', { legs: 4, color: 'brown' })

**Comment by dmarcelino** [comment]
Sounds good! :+1: 

Is the choice of keyword `conditions` based on some other ORM?

**Comment by atiertant** [comment]
@dmarcelino yes cakephp !

**Comment by dmarcelino** [comment]
Wonderful, I've replaced "needs community feedback" :arrow_right: "help wanted" to see if someone wants to volunteer and submit a PR.

**Comment by devinivy** [comment]
In waterline we typically call this "criteria" (rather than "conditions").  I think this is a subfeature of a broader feature to apply default criteria to a collection.  There's a question as to how a user should be able to override the defaults, since in this case you want the criteria to be merged.  I like the feature– let's just spec it out a bit and possibly get a broader feature out of this issue.

**Comment by dmarcelino** [comment]
Regarding the "broader feature", there is already #486 covering "default sort" which we should also consider into the spec.

**Comment by atiertant** [comment]
@devinivy i was not exactly talking about a default criteria for populate but real conditions on the association that could not be override.in the sample i proposed : a dog is a pet with 4 legs,it is not possible to have a dog with 2 legs,specify a criteria of 2 legs in populate whould result searching a pet having 4 legs and having 2 legs,so no results possible.
having default criteras overridable could be useful too but it's not the same.

**Comment by devinivy** [comment]
Ah, I understand the distinction.  So part of this feature would be to enforce the condition when adding or updating associated items as well?  If I try to add a 2-legged dog to a user in that case, it should complain.

**Comment by atiertant** [comment]
@devinivy for sure ! but if user doesn't specify any number of legs ? 
does it would be 4 by default ? what about a ">" or "OR" criteria?
it could be useful to have default values for fields too,so if i add a dog called "boby",it's got automatically 4 legs.

**Comment by devinivy** [comment]
Sure!  Let's just agree on some spec for waterline's behavior before moving forward on this.  I like the idea a lot!

**Comment by eddieajau** [comment]
I would probably handle it like validations are handle in the top-level model. Perhaps something like:

```js
function hasFourLegs(dog) {
    return dog.legs == 4;
}

var User = Waterline.Collection.extend({
  identity: 'user',
  connection: 'local-postgresql',
  attributes: {
    firstName: 'string',
    lastName: 'string',
    dog: {
      collection: 'pet',
      via: 'owners',
      hasFourLegs: true,
      dominant: true
    }
  }
});
```

I _think_ that keeps the methodologies consistent?

**Comment by atiertant** [comment]
@eddieajau this is working for validating new insert but how to find them ?

**Comment by eddieajau** [comment]
Ah I see. In that case I'd rename `conditions` to `where` ??

**Comment by dmarcelino** [comment]
How about a `criteria` / `defaultCriteria` object that could be used at the model level or association level? Example:
```js
function hasFourLegs(dog) {
    return dog.legs == 4;
}

var User = Waterline.Collection.extend({
  identity: 'user',
  connection: 'local-postgresql',

  defaultCriteria : {
    sort: 'firstName ASC'
  }

  attributes: {
    firstName: 'string',
    lastName: 'string',
    dog: {
      collection: 'pet',
      via: 'owners',
      defaultCriteria : {
        hasFourLegs: true
      }
      dominant: true
    }
  }
});
```

This way we could satisfy both this issue and #486 in a consistent manner.

**Comment by atiertant** [comment]
@eddieajau why not.it's more compliant with waterline syntax style.
@dmarcelino function could only work in validation,but not in association criteria... validation should be written as where criteria too,so this should become:

	var User = Waterline.Collection.extend({
	  identity: 'user',
	  connection: 'local-postgresql',

	  defaultCriteria : {
	    sort: 'firstName ASC'
	  }

	  attributes: {
	    firstName: 'string',
	    lastName: 'string',
	    dog: {
	      collection: 'pet',
	      via: 'owners',
	      defaultCriteria : {
	        where: {
	          legs: 4
	        }
	      }
	      dominant: true
	    }
	  }
	});

(defaultCriteria is a bit long,why not defaults ...)

**Comment by dmarcelino** [comment]
> (defaultCriteria is a bit long,why not defaults ...)

Because there is already a [`defaultsTo`](https://github.com/balderdashy/waterline-docs/blob/master/models.md#defaultsto), having both would be confusing to say the least. :smile: If `defaultCriteria` is too long we can perhaps push for `criteria`.

**Comment by atiertant** [comment]
@devinivy what do you think of this ?

	var User = Waterline.Collection.extend({
	  identity: 'user',
	  connection: 'local-postgresql',

	  criteria : {
	    sort: 'firstName ASC'
	  }

	  attributes: {
	    firstName: 'string',
	    lastName: 'string',
	    dog: {
	      collection: 'pet',
	      via: 'owners',
	      criteria : {
	        where: {
	          legs: 4
	        }
	      }
	      dominant: true
	    }
	  }
	});

all criteria passed in find,... or populate will be merged in model or association criteria : 
* in case of where in find: the where criterias are join with a AND operator
* in case of where in create: data will be validate before insert
* in case of sort: model or association criteria sort is the last sort order.

**Comment by devinivy** [comment]
I see what you're saying– yeah, I think that's the right idea.  The most complicated part is probably figuring out where to place the create/update validation step.

**Comment by atiertant** [comment]
this could be done by upgrading models callbacks to handle array like this : 

    beforeCreate: [
      function(values, next) {
        next();
      },
      function(values, next) {
        next();
      }]

to add a behavior function and push the model définition callback after.
behavior could use https://github.com/balderdashy/waterline-criteria to help parsing validation criteria

**Comment by atiertant** [comment]
@devinivy is it ok for a PR ?

**Comment by devinivy** [comment]
I _think_ so.  In this case I would like @tjwebb or @particlebanana to weigh-in.  This has been a relatively small conversation among the four of us.  Do we need more community feedback?  I suppose 0.11 could always make a change to this spec if the community doesn't like how it works.

**Comment by sailsbot** [comment]
Thanks for posting, @atiertant.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
This is an ongoing discussion.

**Comment by sailsbot** [comment]
Thanks for posting, @atiertant.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Reopening until this is PRed into the roadmap.

**Comment by leobudima** [comment]
+1, very useful and nice feature

**Comment by devinivy** [comment]
Now in roadmap.

**Comment by mdemblani** [comment]
I think this should be in the roadmap. It would definitely help if something like this would be there in waterline. Would reduce the default criteria that I would have to put in each and every populate

**Comment by atiertant** [comment]
@mdemblani [associations criteria](https://github.com/Atlantis-Software/offshore-docs/blob/master/models/associations/criteria.md) is working on [offshore](https://github.com/Atlantis-Software/offshore), you could use [sails-hook-orm-offshore](https://github.com/Atlantis-Software/sails-hook-orm-offshore) to integrate it in sails.
